pkg/urbit/noun/imprison.c

```diff
@@ -465,6 +465,37 @@ u3i_vint(u3_noun a)
   }
 }
 
+/* u3i_defcons(): allocate cell for deferred construction.
+**            NB: [hed] and [tel] pointers MUST be filled.
+*/
+u3_cell
+u3i_defcons(u3_noun** hed, u3_noun** tel)
+{
+  u3_noun pro;
+  u3t_on(mal_o);
+
+#ifdef U3_CPU_DEBUG
+  u3R->pro.cel_d++;
+#endif
+
+  {
+    c3_w*     nov_w = u3a_celloc();
+    u3a_cell* nov_u = (void *)nov_w;
+
+    nov_u->mug_w = 0;
+
+    //  XX zero-initialize head and tail?
```

Comment:
Can't hurt, we're about to write to it again relatively soon right? `u3_none` may be more appropriate, not sure what the pitfalls of storing one in memory are.

Reply:
It certainly can't hurt. But can it help? This API is somewhat unsafe compared to `u3i_cell()` (as the function comment says, you *must* stored something in both head and tail pointers). We spend a massive percentage of runtime allocating cells, so any memory stores that we can elide in that codepath seem worth pursuing.

Reply:
I claim the strong form of "can't hurt", that it will not measurably affect performance.

All current callers immediately write to `hed`, and my expectation is two consecutive stores to the same location in L1 are much less expensive than the generic memory store, as are ones to two consecutive words. (The shadowed one is hopefully outright optimized out when inlining, I know gdb has sth like https://godbolt.org/ which could be used to check).

Probably the "correct" course here is to profile how much time the PR saves overall when replaying a representative ship(or w/e fleet sandbox test code), and rerun that both with and without the returning uninitialized memory speedup. Otoh the change in general is obviously beneficial for both speed and readability, so that may be not otherwise be warranted. @philipcmonk thoughts?

Reply:
If you prefill these, it might be nice to fill them with dead beef instead of 0 so improper usage crashes instead of being an implicit null terminated list.

Reply:
Is `u3__none` a reasonable deadbeef, or do we have some other canonical non-`Option`-oriented sentinel?

Reply:
`u3_none` (not-a-noun) is the right sentinel value. I've updated this to initialize to it under U3_MEMORY_DEBUG. It's a bug and misuse of the api to ever miss a store through one of these pointers, and the only reason to initialize them is to help catch such a bug.